[#16] scale-out 설정

dding94 opened the pull request.

dding94 committed:
Merge remote-tracking branch 'origin/feature/11-product'

# Conflicts:
#	src/main/java/com/flab/livecommerce/infrastructure/config/MybatisConfig.java

dding94 committed:
Merge branch 'main' of https://github.com/f-lab-edu/LiveCommerce

dding94 committed:
Merge branch 'main' of https://github.com/f-lab-edu/LiveCommerce

dding94 committed:
feat: redis template 설정

dding94 committed:
feat: redis template 을 사용한 세션 불일치 해결

f-lab-bright commented on src/main/java/com/flab/livecommerce/config/RedisConfig.java:
```diff
@@ -0,0 +1,34 @@
+
+    @Bean
+    public RedisConnectionFactory redisConnectionFactory() {
+        return new LettuceConnectionFactory(redisHost, redisPort);
```
lettuce 외에도 다른 클라이언트들이 있는데, 왜 lettuce를 사용할까요??

dding94 replied:
Jedis 도 있지만 Spring Boot 2.0 부터에서 기본 클라이언트에서  deprecated 되었을 정도로 한계가 있기 때문입니다.(동기로 요청 처리 Lettuce 와 비교하였을 때 TPS, CPU 사용량, 응답속도 차이가 남) 
레디스 홈페이지에도 실려있듯이 확장성이 뛰어난 것이 필요하면 Lettuce 를 사용하라고 권장합니다. 
이유는 Lettuce는 Netty (비동기 이벤트 기반 고성능 네트워크 프레임워크) 기반의 Redis 클라이언트로 Redis에서 기본적으로 지원하는 모든 운영 모델(클러스터, 센티넬 ...)과 연결됩니다. 비동기로 요청을 처리하기 때문에 속도도 월등합니다.

f-lab-bright replied:
두 녀석의 실제 동작 차이도 한 번 찾아서 비교해보시면 좋을 것 같습니다 :)

f-lab-bright commented on src/main/java/com/flab/livecommerce/infrastructure/user/persistence/redis/RedisTokenRepository.java:
```diff
@@ -0,0 +1,27 @@
+        this.redisTemplate = redisTemplate;
+    }
+
+    public void save(String token, User user) {
+        redisTemplate.opsForValue().set(token, user);
+    }
```
User Entity를 그대로 저장하는게 바람직할까요??
session timeout (n초 이상 요청 없을 시 세션 만료되게 해야 한다면..)을 지원한다면 어떻게 할 수 있을까요?

dding94 replied:
엔티티를 그대로 저장하는 방식은 다른 방법을 좀 더 생각해보고 찾아보도록 하겠습니다.
세션 타임아웃은 레디스에 저장시에 만료시간을 설정할 수 있는데 해당방법을 사용하면 될 것 같습니다.

f-lab-bright replied:
만료시간을 지정하는 책임은 어디에 있는게 적절할까요?

dding94 replied:
세션에 사용자의 모든 정보를 저장하게 되면 메모리 사용량도 불필요하게 높아져서 최소한의 정보를 저장하는게 좋을 것 같습니다.
만료시간을 지정하는 책임은 토큰이 가지는게 적절하다고 생각합니다. 사용자에 요청에 토큰이 생성되는데 그 시점에 만료시간을 가져야 하기 때문입니다.

f-lab-bright replied:
해당 세션에 요청이 들어올 때 마다 만료시간이 갱신되어야 한다면요??

dding94 replied:
질문의 의도를 잘 파악하지 못했습니다... 혹시 요청이 들어올 때 마다 만료시간이 갱신되어야 하면 서버에 많은 요청이 가고, 만료시간이 너무길면 서버에 요청은 줄지만 보안상의 위험때문일까요?

f-lab-bright replied:
아뇨 그렇게 해야 하면 만료 시간을 갱신하는 책임이 어디있어야 할지, 어떻게 갱신되록 설계할지 질문드린거에요

dding94 replied:
세션에 저장되는 값으로 유저 아이디(pk)를 사용하지 못하면 대안으로 생각나는게 회원가입시에 유저에게 고유한 랜덤값을 부여하여 사용하여, 주문시에도 아이디가 아닌 해당값을 사용하면 될 것 같은데 적합한 방법이 될까요?

f-lab-bright replied:
세션에 user id는 무조건 들어가있어야 사용자를 식별할 수 있습니다.
대신 entity에 있는 name 같은건 가지고 있을 필요가 없겠죠.
그 외에 session 유지를 위한 다른 정보들이 별개로 저장되어 있어야 할 수도 있구요.
예를 들면 만료시간..?

dding94 replied:
만료시간은 세션유지를 위해 필요한데. 이전 멘토링시간에 주문시에 id 를 사용하는게 적합하냐는 질문이 떠올라서.. 해당 부분도 동일하게 생각한 것 같습니다.

f-lab-bright replied:
그건 request body에 내가 주문할 때 나의 user id를 넣는게 적절하지 않다는 이야기였구요.
내가 주문할 때는 session에서 user id를 꺼내서 쓰라는 이야기였습니다.
나의 행위는 항상 내가 user id를 직접 지정하는게 아니라, 세션의 정보를 통해서 이루어져야 행위 부정이 일어나지 않을 수 있습니다 :)

dding94 replied:
그렇군요 감사합니다!!

dding94 committed:
refactor: Exception BaseException 상속하여 처리하도록 변경

handler 도 각 도메인별이 아니라 common 에서 처리하도록 변경

dding94 committed:
refactor: Mysql -> JdbcTemplate 이름 변경

dding94 committed:
refactor: User Domain 리팩토링

Manger 를 하나로 변경하여 컨트롤러는 하나의 창구를 통하여 로직수행,
Adapter 에서 사용하는 Repository 는 하나로 변경,
LoginProcessor 에서 로그인과 토큰 저장까지 하나의 흐름으로 변경,
infrastructure layer 에서 Application layer 가 아닌 도메인을 의존하도록 변경

dding94 committed:
refactor: 에러코드 추가

dding94 committed:
refactor: 레디스템플릿 세션 만료시간 설정 밑 저장 로직 변경 및 저장 값 수정

f-lab-bright commented on src/main/java/com/flab/livecommerce/domain/user/TokenRepository.java:
```diff
@@ -2,9 +2,9 @@
 
-    void save(String token, User user);
+    void save(String token, String userId);
 
```
리턴되는 문자열에는 무엇이 저장되는건가요??
이 정보로 로그인 되어있는 사용자의 정보를 충분히 파악하고 활용할 수 있을까요??

dding94 replied:
리턴되는 문자열은 유저 id 값으로 해당 id 만 사용하면 될 것 같다는 생각을 했습니다. 
그렇게 관리가 된다면 만료시간이 갱신되는 지점과 이메일이 필요하면 해당 id를 통하여 db에 접근을 하여 유저 정보를 가져오면 어떨까 생각을 했는데 해당 방법은 불필요한 디스크io가 발생하므로 레디스의 장점중 하나가 사라진다고 생각을 합니다.
아래 말씀해주신 내용이 레디스에 저장되는 value 값으로 새로운 유저 정보(세션관리에 필요한 정보만) 별도의 클래스를 만들어 저장시키고 로그인시에 해당 세션에서 값을 꺼내서 사용할테니 해당 값을 꺼내는 시점에 만료시점을 설정하라는 의미일까요?

f-lab-bright replied:
spring security에서의 @AuthenticationPrincipal 을 사용하는 형태도 한 번 참고해보시면 좋겠네요 :)
userId만 쓴다면 이후에 권한이나 역할 검증을 위해서 매번 디비에 질의해야하는 문제가 생기기 때문에, 사용자의 자주 변하지 않지만, 자주 조회하게될 정보는 세션, 토큰에 담아서 사용하는게 여러모로 좋습니다 :)

f-lab-bright commented:
token 과 매핑되는 정보로 새로운 객체를 만들어서 저장하면 어떨까요??
예를들면 AuthenticatedUser 라는 클래스를 만들고, 이 클래스에 로그인 된 사용자의 정보를 담아둘 수 있다면요..?
ex) userId, email, role, expiredAt … 등의 정보를 유지해두고, token에서 해당 정보가 꺼내질 때 expiredAt을 갱신해서 업데이트 해줄 수도 있겠네요 :)

dding94 committed:
refactor: user role 추가 및 redis 토큰 저장관련 로직 수정

value 로 AuthenticatedUser 를 별도의 객체를 통해 조회 가능성이 높은 정보 같이 저장, 레디스에서 정보를 가져올 때 만료시간 갱신되도록 설정

dding94 commented:
AuthenticatedUser 클래스를 만들어 봤는데 문제점이 발생할 수 있는 부분이나 더 생각해야 할 부분이 있는지 확인 부탁드려도 될까요?

그 외로 로그인시도를 계속하면 토큰이 계속 만들어져 레디스에 저장되는 문제가 있는데 해당 부분은 방법이 아직 마땅히 떠오르지 않아 좀 더 생각해보겠습니다. 

레디스에 저장될 때 아래와 같이 직렬화시에 GenericJackson2JsonRedisSerialize를 사용하여 직렬화를 합니다. @class 가 붙어 저장이 되는데 패키지구조가 다 드러나서 좋지 않은 방법같다는 생각합니다. 다르게 좋은 방법일까요? 

레디스와 Inmemory 는 peresistence 패키지안에 있어도 괜찮을지 고민입니다 persistence 의 단어 의미가 영구히 저장된다는 느낌이 있어서. 이 패키지들(redis, inmemory)을 밖으로 두어야 할지, persistence에 넣어야할지 어떤게 좀 더 괜찮을까요 

![스크린샷 2022-10-22 오후 6 40 47](https://user-images.githubusercontent.com/21376853/197332551-af2fb56c-758d-4bf4-9edf-958bec68d603.png)

f-lab-bright commented:
- 여러 디바이스에서 중복적으로 로그인을 할 수 있게 하려면, 여러 사용자가 로그인 할 때 마다 새 토큰이 저장되는 동작이 맞습니다. 다만 최대 중복 로그인 갯수를 제한 한다면 저걸 검사할 수 있는 구조를 생각해보시긴 하셔야겠네요 :)
- redis를 사용할 때의 serialization의 기본동작이긴 한데, 큰 문제는 없습니다. 다만 패키지 구조가 변경되거나 했을 때는 동작에 문제가 생길 수는 있습니다. 다른 자료구조를 이용하실 수도 있겠네요 :)
- persistence는 꼭 영구히 저장되는 것들을 관리하는 곳은 아닙니다. 저장과 관련된 부분을 처리하는 것들 모두 persistence layer라고 생각하셔도 됩니다. 영구적인 저장소와 휘발성 저장소를 구분하시는게 보기 좋을 것 같다면 그렇게 하셔도 됩니다 :)

f-lab-bright commented on src/main/java/com/flab/livecommerce/common/AuthenticatedUser.java:
```diff
@@ -0,0 +1,24 @@
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+public class AuthenticatedUser {
+
+    public static final long EXPIRE_TIME = 1000L;
```
유효기간을 얼마나 사용할건지는 설정으로 변경할 수 있게 되는게 좀 더 바람직할 것 같아서, 이 클래스 내에 static variable로 가지고 있는 것 보다는 외부에서 값을 변경하거나 지정할 수 있게 구성하는게 좀 더 유연할 것 같네요 :)

dding94 replied:
이 부분이 잘 이해가 안되는데 외부에서 값을 변경하거나 저장할 수 있게 구성한다는 것이 클라이언트 요청으로 유지시간이 변경되는게 아니라면 common 패키지에 Util 클래스로 다양한 만료시간값을 가질 수 있도록 하는게 괜찮을까요?

f-lab-bright replied:
서버의 설정만으로 만료시간을 변경할 수 있도록 구성해보라는 이야기입니다 :)
지금 코드라면 만료시간을 변경하려면 코드를 변경하고 재빌드 후 재배포를 해야 하니까요 :)

dding94 replied:
아 그렇군요 서버의 설정값을 활용하면 변경시의 재빌드의 필요가 없어지겠군요 감사합니다.

f-lab-bright commented on src/main/java/com/flab/livecommerce/common/AuthenticatedUser.java:
```diff
@@ -0,0 +1,24 @@
+    private Long userId;
+    private String email;
+    private Role role;
+    private Long expireAt;
```
Long 타입으로 expireAt으로 지정하면 이게 어떤 scale의 값인지, 어떻게 사용될지 이해하기 어렵겠네요.

dding94 replied:
처음에 해당 부분을 고려할 때 만료시간을 가지는 책임이 해당 클래스에 있는줄 알고 Long 타입으로 설계했습니다.
만료되는 시간을 계산하여 날짜를 표기하여 저장할 수 있도록 LocalDateTime 을 사용하도록 변경했습니다.

f-lab-bright commented on src/main/java/com/flab/livecommerce/infrastructure/user/TokenRepositoryAdapter.java:
```diff
@@ -1,31 +1,32 @@
     @Override
     public void save(String token, User user) {
-        inMemoryTokenRepository.save(token, user);
+        tokenRepository.save(token, user);
```
저장할 떄는 user로 저장하는데, 꺼내올 때는 AuthenticatedUser로 꺼내올 수 있나요?

f-lab-bright commented on src/main/java/com/flab/livecommerce/infrastructure/user/redis/RedisTokenRepository.java:
```diff
@@ -0,0 +1,51 @@
+        this.redisTemplate = redisTemplate;
+    }
+
+    public void save(String token, User user) {
+        redisTemplate.opsForValue().set(
+            token,
+            new AuthenticatedUser(
+                user.getId(),
+                user.getEmail(),
+                user.getRole()
+            )
+        );
+        expireTimeUpdate(token);
+    }
```
User를 저장할 때 내부적으로 변환해서 저장하는 것 보다는, 애초에 interface가 변환된 객체를 저장할 수 있게 하는게 좀 더 가독성이 좋은 구조가 되지 않을까요?

dding94 replied:
내부적에서 처리를 하지 않는 것이 해당 부분에서 new 연산자를 사용하지 않고 만드는 것이 가독성 측면에서도 좋은 구조가 될 것 같습니다. 위에서 말씀한것을 더해  저장과 꺼내오는 메서드를 동일하게 변경하였습니다.

f-lab-bright reviewed (commented):
AutenticatedUser 클래스로 저장할 뿐이고, 이를 활용하는 부분이 사실상 거의 없어보이네요 :)
Controller의 api mapping method에서 인증이 필요한 메소드라면 AuthenticatedUser를 파라미터로 받을 수 있게 구성할 수 있으면 좋을 것 같네요 :)

f-lab-bright commented on src/main/java/com/flab/livecommerce/infrastructure/user/redis/RedisTokenRepository.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    public void save(String token, User user) {
+        redisTemplate.opsForValue().set(
+            token,
+            new AuthenticatedUser(
+                user.getId(),
+                user.getEmail(),
+                user.getRole()
+            )
+        );
+        expireTimeUpdate(token);
```
이렇게 하면 redis에 두 번 요청을 보낼 것 같은데, 확인해보셨나요??
만약 실제로 두 번 호출이 된다면 한 번으로 줄일 수는 없을까요??
rtt가 두 번 일어나는걸 한 번으로 줄이는건 성능상 많은 차이가 생깁니다 :)

dding94 replied:
멘토님 호출에 대한 부분은 만료시간을 함께설정하는 메서드로 한번으로 줄일 수 있으나, 해당 정보가 꺼내질 때 expiredAt을 갱신해서 업데이트 해준다고 말씀을 하셔서 이 부분에 대한 요청이 이뤄져야 할 것 같아,  토큰으로 조회시에 갱신까지 이루어진다면 조회1번 저장1번 이렇게 2번 요청을 보내는것은 상관이 없을까요?

f-lab-bright replied:
직접 한 번 설계하고 구현해보시죠 :)
한 번 호출해도 될 부분을 두 번 호출하지만 않으면 괜찮습니다 :)

dding94 committed:
refactor: TokenRepository, RedisTokenRepository 수정

save, find (return) 은 같은 AuthenticatedUser 객체 사용하도록 변경, 조회시에 불필요한 호출 제거, 가독성을 위하여 AuthenticatedUser 내부에 객체생성 메서드 추가

dding94 committed:
chore: common 으로 패키지 위치 변경

user 도메인에서만 적용되는 것이 아니기 때문

dding94 committed:
test: user 도메인 테스트코드 작성

로그인, 회원가입 테스트

f-lab-bright commented:
컨플릭트도 해결해주세요 :)

dding94 commented:
알겠습니다!!

f-lab-bright commented on src/main/java/com/flab/livecommerce/application/user/UserLoginProcessor.java:
```diff
@@ -42,7 +43,7 @@ public String execute(LoginCommand command) {
 
         var token = tokenGenerator.generate();
-        tokenRepository.save(token, user);
+        tokenRepository.save(token, AuthenticatedUser.create(user));
```
token도 user entity처럼 authenticatedUser의 unique id 처럼 클래스 안의 필드로 넣어두고 사용해도 괜찮지 않을까요??

f-lab-bright commented on src/main/java/com/flab/livecommerce/common/AuthenticatedUser.java:
```diff
@@ -8,17 +10,28 @@
-        this.expireAt = EXPIRE_TIME;
+    }
+
+    public AuthenticatedUser updateExpireTime(long time) {
```
updateExpireTime 이라는 이름이면 파라미터의 값으로 치환하는 것 처럼 읽힐 것 같네요! :)
그리고 seconds만큼 추가하는 거라면 time 같이 scale을 알 수 없는 명칭보다는 secs, seconds, s 등 의 파라미터를 어떤 값으로 넣어야 하는지 좀 더 이해하기 편하게 선언하시는게 가독성 측면에서 더욱 좋습니다 :)

dding94 replied:
다른 누군가 본다면 착각하고, 혼동하기 쉬운 언어로 사용을 했었네요
메서드명이나 파라미터명을 누가 읽어도 파악할 수 있게 구체적으로 작성하도록 하겠습니다!

dding94 committed:
Merge branch 'main' into feature/16-scale-out

f-lab-bright commented on src/main/java/com/flab/livecommerce/config/RedisConfig.java:
```diff
@@ -18,17 +22,26 @@ public class RedisConfig {
     private int redisPort;
 
+
+    @Bean
+    public ObjectMapper objectMapper() {
+        ObjectMapper mapper = new ObjectMapper();
+        mapper.disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);
+        mapper.registerModules(new JavaTimeModule(), new Jdk8Module());
+        return mapper;
+    }
```
spring에서 기본적으로 objectMapper가 이미 빈으로 등록되어 있는 녀석이 있는데, 새로 만들어서 사용하시는 이유가 있으신가요??
스프링이 등록해주는 objectMapper 빈과는 다른 설정으로 사용해야 할게 있나요??
objectMapper를 request / response 시에 dto의 serialization / deserialization 동작에 영향을 미치는 빈이라 주의하셔서 추가 등록을 하셔야 될 수 있습니다 :)

dding94 replied:
이미 objectMapper 가 빈으로 등록되어 있다는 사실은 알지 못했습니다. 만든 이유는 authenticatedUser 객체를 json 변환시에 날짜와 Optional 타입 매핑을 위해 만들어서 사용하려고 그렇습니다. 기존에 등록되어 있는 objectMapper를 살펴보고 현재 등록된 빈 이름을 변경하던지 다르게 사용할 수 있는 방법을 알아보도록 하겠습니다.

dding94 replied:
알아보니 현재상황에서는 기본으로 등록하는 mapper 와 다른 설정으로 사용될만한게 없네요.. 사용자 지정 날짜 형식을 갖도록 module 을 따로 만들어서 사용하는것도 아니기 때문입니다.

f-lab-bright replied:
네 :) 그리고 objectMapper가 thread safe한 녀석이라 기존의 녀석을 사용해도 전혀 문제 없기도 하구요 :)
만약 특별한 설정의 objectMapper를 사용하고 싶으시다면, 그리고 그 녀석을 여러 빈에서 사용하려는게 아니라면 bean으로 등록하시기 보다, 구현 클래스 내에서 직접 생성해두고, 필드, 혹은 스태틱 필드로 가지고 있도록 해서 사용하시는게 좀 더 안전할 수도 있습니다 :)

dding94 committed:
Merge branch 'main' of https://github.com/f-lab-edu/LiveCommerce

dding94 committed:
Merge branch 'main' into feature/16-scale-out

# Conflicts:
#	src/main/resources/application.yml

f-lab-bright commented on src/main/java/com/flab/livecommerce/infrastructure/user/persistence/redis/RedisTokenRepository.java:
```diff
@@ -0,0 +1,58 @@
+    private final RedisTemplate<String, Object> redisTemplate;
+    private final ObjectMapper objectMapper;
+
+    @Value("${expire.default}")
+    private long expireTime;
```
이 값을 가지고 있는 녀석이 repository 이어야 할지도 고민해보면 좋을 것 같고, 테스트를 위해서 이 값도 @Value로 바로 필드 인젝션하기보다는, constructor로 주입받을 수 있게 하시는게 좋겠네요! :)
@Bean 으로 등록할 때 @Value() 어노테이션으로 옵션을 읽어오시는게 좀 더 좋겠네요 :)

dding94 committed:
refactor: redis config 수정

ObjectMapper 를 빈으로 따로 만들어서 등록하게 되는 경우는 스프링에서 기본적으로 등록해주는 ObjectMapper 와 충돌의 위험이 있음

dding94 committed:
feat: 세션 로그인 정보가져오는 기능 추가(argument resolver 활용), 수정필요한 부분 남음

패키지 구조 변경, 임시 테스트 컨트롤러 추가, 인터셉터에서 헤더에 세션정보 담을 수 있도록 변경, 인증 에러코드 추가, expire 관련 AuthenticatedUser, redis 로직은 수정이 필요

dding94 committed:
fix: 컨벤션 체크, import 위치 변경

dding94 committed:
refactor: redis 세션 리팩토링

expire 시간을 가지는 TokenProperties 클래스 생성, 인터셉터 책임 분리 (로그인, 세션) 및 WebConfig 등록, 로그인프로세서 로직 수정(인증 유저 객체 생성시 만료값 주입 받도록), TokenRepository 만료시간 갱신 메서드 추가,

dding94 committed:
test: 더미나 fake 만들어서 테스트 할 수 있도록 변경해야함

f-lab-bright commented:
@DJ-archive 요거 다정님도 리뷰 해보시고, 문제 없으면 approve 하셔서 병합해보시죠 :)

dding94 committed:
chore: jpa 의존성 추가

dding94 committed:
test: jpa persistence-context test

테스트 전용 docker 추가, jpa 영속성 컨텍스트 테스트
Ci 체크 통과하는 방법을 몰라서 주석처리 함.
- 추후 삭제하던지, 통과하도록 변경 (도커 포함)

2eve2 commented on src/main/java/com/flab/livecommerce/common/Role.java:
```diff
@@ -0,0 +1,5 @@
+package com.flab.livecommerce.common;
```
Role 클래스를 Domain-User가 아닌 common에 위치시킨 이유가 있으실까요?

dding94 replied:
partner 에도 Role 클래스를 사용해야하지 않을까요?
공통으로 사용되는 부분이라 common 에 위치시켰습니다. 좋은 방법 있으시면 남겨주시면 감사하겠습니다.

2eve2 replied:
아하 partner도 애플리케이션 사용자에 해당하니까 user 도메인에 위치하는 것이 맞다고 생각했었습니다.
말씀 들어보니 명구님은 일반 user랑 partner랑 아예 다른 도메인으로 분류하려고 하셨던 거였네요. 저는 User 도메인 안에서 역할만 나눠지는걸로 생각하고 있었습니다. 

shop 도메인이 별도로 email, business number 등 상태 정보가 존재하기에 shop으로 입점(회원가입)을 하고 user를 저장할 때 Role의 partner로 저장하는 방식으로 생각했었습니다.

dding94 replied:
예 저번에 말씀드린걸로 기억을 하는데 제대로 의견전달이 안됐나 봅니다.
로그인 자체가 사업자랑 일반 사용자랑 나뉘는 경우를 더 많이 본 것 같아서 그렇게 생각하고 도메인을 분리를 하였습니다.

2eve2 replied:
shop 도메인에서 email, password 등을 다 만드셨던데, partner를 따로 또 도메인으로 만들어서 가입하게 할 생각이신건가요?
위에 말씀드렸듯이 저도 따로 분리해서 회원가입하는건 이해했었으나, 다만 shop으로 가입한다고 생각했습니다. 
지금 객체 명칭이 달라서 헷갈리는데 shop과 partner는 어떻게 다른건지 설명해주시면 감사드리겠습니다.

2eve2 replied:
같은 도메인이라면, 기존의 'shop' 명칭으로 Role명도 통일하는 것이 좋을 것 같습니다.

dding94 replied:
그래서 처음에 partner 라고 도메인명을 지어서 만들었는데 다정님이 shop 으로 변경하자고 요청을 하셔서 변경했습니다. 
오히려 제가 shop 과 partner 가 어떻게 다른건지 설명을 요청드리고싶네요..

2eve2 replied:
아 네.. 제가 헷갈린 것을 말씀드리자면..

shop으로 명칭 통일하게끔 합의한 상황에서 partner를 별도로 만드시니까 개인인 partner가 여러개의 shop을 가지도록 설계하시는건가하고 생각하게 되었습니다. 이렇게되면 User 도메인 안에 Role이 들어가도 되니까요. 

같은 도메인을 나타내고자 하신거라면, 처음에 명칭 협의한대로 shop으로 통일하는 것이 맞다고 생각합니다. 다른 이의 코드를 보는 입장에서 헷갈릴 여지가 있다고 생각합니다..ㅠ

사실 별 거 아니라고 느낄 수 있지만.. 오해가 생길까봐 말씀드리는데 정말 순간 헷갈렸습니다.

dding94 replied:
아 그렇군요. 그러면 shop 을 partner 라고 바꾸는건 어떨까요? seller 도 될 수 있겠네요.
저는 해당부분 만들면서 shop 이라는 명칭이 과연 맞나 이생각이 들었습니다.

2eve2 replied:
아 넵 shop과 partner 둘다 서로 맞지 않다고 느끼니, seller로 통일할까요?

dding94 replied:
다정님께서 seller 괜찮으시면 해당 명칭으로 가시죠!

DJ-archive commented:
현재 Controller API 상에서 로그인 체크가 적용되어있지 않은 api는 다른 pr에서 작업 진행할 생각이실까요?
partner 외 user 인증, 인가 스케일 아웃 파트를 작업하셨는데, test api를 별도로 만드셔서 여쭤봅니다. 
가령 현재 구현된 Order register api를 user id로 @logincheck 하게끔 수정하는 것이 있겠습니다.

2eve2 commented on src/main/java/com/flab/livecommerce/presentation/user/UserController.java:
```diff
@@ -55,4 +51,10 @@ public CommonApiResponse checkEmail(@RequestBody @Valid UserEmailRequest email)
         return CommonApiResponse.success(null);
     }
+
+    @LoginCheck
+    @PostMapping("/test")
+    public CommonApiResponse test(@Authentication AuthenticatedUser authenticatedUser) {
+        return CommonApiResponse.success(authenticatedUser);
+    }
```
현재 Controller API 상에서 로그인 체크가 적용되어있지 않은 api는 다른 pr에서 작업 진행할 생각이실까요?
partner 외 user 인증, 인가 스케일 아웃 파트를 작업하셨는데, test api를 별도로 만드시고 기존 api 수정은 없어 여쭤봅니다.
가령 현재 구현된 Order register api를 user id로 @logincheck 하게끔 수정하는 것이 있겠습니다.

dding94 replied:
일단 첫번째에서 말씀해주신 부분은 다른 pr 에서 작업할거고 test api 도 지울 예정입니다.
이렇게 만들었다라는 것을 멘토님과 다정님에게 보여주려고 한 이유도 있습니다.
마지막에 말씀해주신 Order register api 를 user id 로 @logincheck 하게끔 수정한다는게 무슨 말씀인지 잘 모르겠습니다.
제가 착각하고 있는 부분이 있을까 여쭙습니다. 로그인 체크는 @logincheck 어노테이션만 사용하면 되도록 만든게 아니였을까요?
저는 user id 는 HttpSession 에서 값을 읽어오듯이 redis 에서 정보를 읽어와 AuthenticatedUser 에서 Id 값을 가져오도록 설계를 했습니다.

2eve2 replied:
아 뒷 내용에 대해 제가 말 전달을 잘 못한 것 같습니다. 그리고 확인해보니 테스트 api 역할과 redis에 대해 착각하고 있었네요.

잘못 전달한 내용은, order register api에서 user id가 requestDto로 전달받고 있는데 이를 pathvariable로 받게끔 수정해서 데이터 변조 위험을 막은 api로 수정해야 한다. 그 후 @logincheck와 redis 방식 적용해야 한다는 뜻이었습니다. 하지만 해당 pr에서는 굳이 적용시키지 않아도 되는 내용이며, redis 부분을 잘못 이해하였고 있었습니다. 

그리고 혹시 앞으로 api 이름 자체로 이해하기 어려운 경우, 위에 주석으로 설명이나 uri을 좀 더 자세히 적어주실 수 있을까요? 가령 로그인 세션 정보를 보는 api라는 설명을 적거나, uri 주소를 test의 주체를 명확히 해서 변경하면 좀 더 이해가 쉬울 것 같습니다.

그리고 @Authentication 어노테이션 사용 코드를 보니까 parameter가 제대로 된 AuthenticatedUser가 맞는지 확인하는 용도로 사용된 것 같은데 제가 이해한 것이 맞을까요?

dding94 replied:
해당 api 는 삭제할 용도로 만든거라 메서드명이나 주석이나 하나도 고려를 하지 않았습니다.
이렇게 만들고 지우게 될 api 를 남기게 된다면 요청하신대로 주석을 달아 두겠습니다.

넵 해당 어노테이션이 붙은 객체는 AuthenticatedUser 클래스인지 확인합니다. 
그리고 따로 구현한 AuthenticationArgumentResolver 를 실행할지 판단을 합니다.

2eve2 replied:
알겠습니다.
그렇다면 Authentication 이름의 어노테이션보다 다른 명칭을 쓰는건 어떻게 생각하시나요? AuthenticatedUser 클래스인지 확인하는 어노테이션인데, 인증이라는 의미가 들어가서 로그인 인증과 의미가 헷갈렸습니다.

dding94 replied:
헷갈리시다면 어떤 명칭을 사용하시길 원하시나요? 인증된 사용자 객체를 만들지 판단하는 어노테이션인데 지금당장 마땅히 떠오르는것은 없네요. 이 어노테이션이 붙은 객체는 AuthenticatedUser 으로 고정되어있고 개인적인 의견은 합리적이라고 생각됩니다.

dding94 replied:
아니면 지금 LoginUser 어노테이션이랑 바꾸는것도 방법이지 않을까 생각이 드네요.

2eve2 replied:
당장은 @AuthenticatedUser, @CheckAuthUser 등이 생각나네요. 

처음 명칭을 보고 이해할 때 기존에 다른식으로 사용했었던 단어로 그대로 보면, 코드를 뜯어보기 전에는 오해가 생기는 것 같습니다. 지금은 저도 어노테이션에 붙을 객체와 의미를 이해했으니 그냥 두셔도 될 것 같습니다!

저도 앞으로 코드를 보면서 제가 알고있는 것과 좀 다르게 작동되는 것 같으면 더 자세히 보도록 하겠습니다.
